### PR TITLE
Closes #223 | Add/Update Dataloader VinText

### LINKED ISSUE #223: Create dataset loader for VinText
Dataloader name: `vintext/vintext.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?vintext

|     Dataset| vintext |
|-------------|---|
| Description | Vintext is a challenging scene text dataset for Vietnamese, where some characters are equivocal in the visual form due to accent symbols. This dataset contains 2000 fully annotated images with 56,084 text instances. Each text instance is delineated by a quadrilateral bounding box and associated with the ground truth sequence of characters. The dataset is randomly split into three subsets for training (1,200 images), validation (300 images), and testing (500 images). |
| Subsets | - |
| Languages | vie |
| Tasks | Optical Character Recognition |
| License | GNU Affero General Public License v3.0 (agpl-3.0) |
| Homepage | https://github.com/VinAIResearch/dict-guided |
| HF URL | - |
| Paper URL | https://ieeexplore.ieee.org/document/9577624 |

### PR DESCRIPTION
**Pull request by patrickamadeus**
Closes #223 

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/{my_dataset}/{my_dataset}.py` (please use only lowercase and underscore for dataset folder naming, as mentioned in dataset issue) and its `__init__.py` within `{my_dataset}` folder.
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_LOCAL`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py` or `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py --subset_id {subset_name_without_source_or_seacrowd_suffix}`.
- [.] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### Tests
image

### CODE REVIEW

File: seacrowd/sea_datasets/vintext/vintext.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            features=features,
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager: datasets.DownloadManager) -> List[datasets.SplitGenerator]:
+        """Returns SplitGenerators."""
+
+        zip_filepath = os.path.join(os.path.dirname(__file__), "vietnamese_original.zip")
```
Comment (akhdanfadh) [inline seacrowd/sea_datasets/vintext/vintext.py:130]: Clever! I never thought that one can get HF's `.cache` folder like this. 😄
I guess you know already that there is a caveat on using cache folder though: once you edit the dataloader file (even a slight edit), the cache dir will be different, thus downloading the dataset again.

I personally store downloaded `gdown` data in a newly-created `data` directory in seacrowd root folder, though there are pros and cons.

File: seacrowd/sea_datasets/vintext/vintext.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                }
+
+        elif self.config.schema == "seacrowd_imtext":
+            for i, row in enumerate(df_list):
+                labels = [label.split(",") for label in row["label"].split("\n")]
+
+                yield i, {
+                    "id": row["id"],
+                    "image_paths": [row["image_path"]],
+                    "texts": "",
```
Comment (akhdanfadh) [inline seacrowd/sea_datasets/vintext/vintext.py:195]: Let's use `None` instead since an empty string is still a string.

```suggestion
                    "texts": None,
```

File: seacrowd/sea_datasets/vintext/vintext.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    booktitle = {Proceedings of the {IEEE} Conference on Computer Vision and Pattern Recognition (CVPR)},
+}
+"""
+
+_DATASETNAME = "vintext"
+
+_DESCRIPTION = """\
+Vintext is a challenging scene text dataset for Vietnamese, where some characters are equivocal in the visual form due to accent symbols.
+This dataset contains 2000 fully annotated images with 56,084 text instances. Each text instance is delineated by a quadrilateral bounding box and associated with the ground truth sequence of characters.
+The dataset is randomly split into three subsets for training (1,200 images), validation (300 images), and testing (500 images).
```
Comment (akhdanfadh) [inline seacrowd/sea_datasets/vintext/vintext.py:46]: This split description is for the converted dataset (in COCO format), yet here you only implemented the original dataset variant. Any reason why don't you implement the converted dataset as well (maybe as a subset)? If it is difficult, I suggest replacing this description with some description saying that this dataloader only implements the original data from the homepage.

File: seacrowd/sea_datasets/vintext/vintext.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            datasets.SplitGenerator(
+                name=datasets.Split.TEST,
+                gen_kwargs={
+                    "imagepath": os.path.join(data_dir, "vietnamese/test_image"),
+                    "labelpath": os.path.join(data_dir, "vietnamese/labels"),
+                },
+            ),
+        ]
+
+    def _generate_examples(self, imagepath: Path, labelpath: Path) -> Tuple[int, Dict]:
```
Comment (akhdanfadh) [inline seacrowd/sea_datasets/vintext/vintext.py:153]: I think `os.path.join` output type is a string, no? So this one should be changed even though there are no errors.

File: seacrowd/sea_datasets/vintext/vintext.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            for i, row in enumerate(df_list):
+                labels = [label.split(",") for label in row["label"].split("\n")]
+
+                yield i, {
+                    "id": row["id"],
+                    "image_paths": [row["image_path"]],
+                    "texts": "",
+                    "metadata": {
+                        "context": "",
+                        "labels": [],
```
Comment (akhdanfadh) [inline seacrowd/sea_datasets/vintext/vintext.py:198]:
```suggestion
                        "context": None,
                        "labels": None,
```

**Review by akhdanfadh — CHANGES_REQUESTED** [review]
Dataloader tested and worked at first, but please address my discussions below.

File: seacrowd/utils/schemas/image_text.py [hunk 6]
```diff
@@ -12,7 +12,18 @@ def features(label_names = ["Yes", "No"]):
...
+                    "y1": datasets.Value("int32"),
+                    "x2": datasets.Value("int32"),
+                    "y2": datasets.Value("int32"),
+                    "x3": datasets.Value("int32"),
+                    "y3": datasets.Value("int32"),
+                    "x4": datasets.Value("int32"),
+                    "y4": datasets.Value("int32"),
+                    "transcript": datasets.Value("string"),
+                })
+            },
```
[the remark this answers is not in the document]
Comment (patrickamadeus) [reply seacrowd/utils/schemas/image_text.py:26]: Initially I thought abt that too, but this implementation has been discussed in discord discussion [thread](https://discord.com/channels/1164061531227557920/1227907459897757747). Wdyt guys? @akhdanfadh @holylovenia

File: seacrowd/sea_datasets/vintext/vintext.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+                            "y4": datasets.Value("int32"),
+                            "transcript": datasets.Value("string"),
+                        }
+                    ),
+                }
+            )
+
+        elif self.config.schema == "seacrowd_imtext":
+            features = schemas.image_text_features()
+
```
[the remark this answers is not in the document]
Comment (patrickamadeus) [reply seacrowd/sea_datasets/vintext/vintext.py:121]: _(same with above's discussion)_

File: seacrowd/sea_datasets/vintext/vintext.py (hunk 1 above)
Comment (patrickamadeus) [reply seacrowd/sea_datasets/vintext/vintext.py:133]: Thank you ahahaha 🙏.

My only concern is I am trying to minimize adding new folder to the SEACrowd repo that does not follow the template. But both is possible though!

The other consideration is, I think the hassle is only from the developer side (the slight change new cache problem). I don't think if one uses our dataset through `load_dataset` the cache will reset (ofc because there is no change to the dataloader file).

Again, please lmk if the `data` directory approach is cleaner! I would be happy to do it. Wdyt guys? @akhdanfadh @holylovenia @fhudi

File: seacrowd/sea_datasets/vintext/vintext.py (hunk 3 above)
Comment (patrickamadeus) [reply seacrowd/sea_datasets/vintext/vintext.py:49]: Thank you for the suggesstion! It's really a surprise most of the reviews were my contemplations also 😂.

Hmmm, my only thinking was it's better to choose one of the format since new `subset` by definition (please correct me if i'm wrong) often has either different content / content & task / content & task & schema within the new `subset`'s dataset. (like `movie_a`, `movie_b`, and `movie_c` type of stuff, they have different content), whereas the only different between `COCO` and `original` subset is only the format (although `COCO` is obviously more detailed).

And the more crucial one is, both of the `subset` will have their very own `SEACrowd` (`seacrowd_coco`, `seacrowd_ori`, etc.) schema, which if we follow current implementation, will be exactly the same (only mapping the `x1, x2,..., xn` from rich `COCO` format). I think it's better to discuss whether to switch `image_text` schema 's format first with @holylovenia if we want to incorporate the `COCO` format.

I'll narrow down the description for now. (even if we proceed with `COCO`, I think it's better to left out the `original`? since the `original` is actually the subset / more concise version of `COCO` format)

Comment (patrickamadeus) [reply seacrowd/sea_datasets/vintext/vintext.py:48]: Well I guess i'm overcomplicating the definition of `subset`. We can implement the `COCO` format too!

Maybe let's wait for the `image_text` second opinion from @holylovenia.

File: seacrowd/utils/schemas/image_text.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fhudi) [reply seacrowd/utils/schemas/image_text.py:26]: @patrickamadeus noted.
I do agree with the thread, i.e. for image & text dataset we should use this format.
However, does not this change (i.e. making the schemas having bounding box annotations) necessitate the change of other datasets without the annotations?
cc: @akhdanfadh @holylovenia

[the remark this answers is not in the document]
Comment (holylovenia) [reply seacrowd/utils/schemas/image_text.py:26]: @patrickamadeus The `annotations` is supposed to be a key in the `metadata` according to our discussion. It was some time ago though, so it makes sense if we forget about this little detail.

image

File: seacrowd/sea_datasets/vintext/vintext.py (hunk 1 above)
Comment (holylovenia) [reply seacrowd/sea_datasets/vintext/vintext.py:136]: @patrickamadeus Either implementation should be fine as long as it works. Let's keep the current implementation since it has been done.

File: seacrowd/sea_datasets/vintext/vintext.py (hunk 3 above)
Comment (holylovenia) [reply seacrowd/sea_datasets/vintext/vintext.py:48]: @patrickamadeus Thanks, I have replied on the comment thread above. 👍

File: seacrowd/utils/schemas/image_text.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fhudi) [reply seacrowd/utils/schemas/image_text.py:26]: Can I ask something, since we are not supporting the task of Scene Text Recognition, why do we _need_ the BB in the metadata?

As this is formulated as an OCR task, _not STR or text localisation_, I would suggest it is OK not to _force_ BB labels in the SEACrowd schema. For 2 reasons:

1. The BB is accessible from the source schema anyway.
2. There is no need to change the `utils/schemas/image_text.py`, but put the BB labels in this particular dataloader only, i.e.: only in the `sea_datasets/vintext/vintext.py`.

[the remark this answers is not in the document]
Comment (holylovenia) [reply seacrowd/utils/schemas/image_text.py:26]: @fhudi `metadata` already exists in the current [`imtext` schema](https://github.com/SEACrowd/seacrowd-datahub/blob/master/seacrowd/utils/schemas/image_text.py). Does adding the `annotations` inside the `metadata` trigger an error for the unit test?

[the remark this answers is not in the document]
Comment (holylovenia) [reply seacrowd/utils/schemas/image_text.py:26]: @fhudi The purpose is precisely to store all the additional information such as bbox. That's intentional. It's good that it doesn't trigger any error, let's keep the `annotations` inside the `metadata` then.

cc: @patrickamadeus

[the remark this answers is not in the document]
Comment (fhudi) [reply seacrowd/utils/schemas/image_text.py:26]: in that case, can we put some comments on the schema file indicating some keys may not be available depending on the dataloader? 🙏

[the remark this answers is not in the document]
Comment (holylovenia) [reply seacrowd/utils/schemas/image_text.py:26]: @fhudi Done.

[the remark this answers is not in the document]
Comment (patrickamadeus) [reply seacrowd/utils/schemas/image_text.py:26]: Doing it rn!

**Comment by patrickamadeus** [comment]
Hi @akhdanfadh! I've addressed the suggestions! For now I will change the description and proceed with the original version of dataset.

Thank you for the meaningful reviews!

**Comment by sabilmakbar** [comment]
Hi, @akhdanfadh @fhudi. Would you like to review the latest code changes? Thanks!

**Comment by holylovenia** [comment]
Hi @fhudi @akhdanfadh, I would like to let you know that we plan to finalize the calculation of the open contributions (e.g., dataloader implementations) **in 31 hours**, so it'd be great if we could wrap up the reviewing and merge this PR before then.

cc: @patrickamadeus

**Review by fhudi — APPROVED** [review]
LGTM.
Passed the tests.
Thanks for being cooperative!

**Comment by fhudi** [comment]
Hi @akhdanfadh,
It seems all the previous concerns have been addressed.
Could you please re-review the latest code changes? thanks.
cc: @sabilmakbar

**Comment by akhdanfadh** [comment]
@fhudi The code tested and OK. My only comment is that the change in `image_text.py` schema is insignificant. Could we remove that?

**Comment by fhudi** [comment]
@akhdanfadh sure, please do.
I am fine with either one.

**Review by akhdanfadh — APPROVED** [review]

**Comment by akhdanfadh** [comment]
Merging now. @patrickamadeus @fhudi

File: seacrowd/utils/schemas/image_text.py
Comment (fhudi) [inline seacrowd/utils/schemas/image_text.py:1]: Thanks for adding additional comment here.
Though, too bad that the new line at the EOF is gone 🥲
